### PR TITLE
Add string type

### PR DESCRIPTION
**Pull request by hekota**
Add string type and enable top level static const declaration of strings in preparation for future features. Static const is always implied. Arrays, vectors and matrices of string are prohibited. Strings in function
parameters and bodies are currently not supported either.

### CODE REVIEW

File: tools/clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 1]
```diff
@@ -7650,6 +7650,8 @@ def warn_hlsl_effect_object : Warning <
 def warn_hlsl_unused_call : Warning<
   "ignoring return value of function that only reads data">,
   InGroup<UnusedValue>;
+def err_hlsl_string_not_global : Error<
+   "string declaration not supported here">;
```
Comment (tex3d) [inline tools/clang/include/clang/Basic/DiagnosticSemaKinds.td]: Perhaps specify where they are supported.

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp [hunk 2]
```diff
@@ -1043,6 +1051,9 @@ unsigned CGMSHLSLRuntime::AddTypeAnnotation(QualType Ty,
     AddTypeAnnotation(GetHLSLResourceResultType(Ty), dxilTypeSys, arrayEltSize);
     // Resource don't count for cbuffer size.
     return 0;
+  } else if (IsStringType(Ty)) {
```
Comment (tex3d) [inline tools/clang/lib/CodeGen/CGHLSLMS.cpp]: If this case shouldn't happen, perhaps replace with DXASSERT.

File: tools/clang/lib/AST/Expr.cpp [hunk 3]
```diff
@@ -833,7 +833,7 @@ StringLiteral *StringLiteral::Create(const ASTContext &C, StringRef Str,
                                      StringKind Kind, bool Pascal, QualType Ty,
                                      const SourceLocation *Loc,
                                      unsigned NumStrs) {
-  assert(C.getAsConstantArrayType(Ty) &&
+  assert(hlsl::IsStringLiteralType(Ty) &&
```
Comment (tex3d) [inline tools/clang/lib/AST/Expr.cpp]: perhaps use DXASSERT for our asserts.

File: tools/clang/lib/Sema/SemaDecl.cpp [hunk 4]
```diff
@@ -10266,6 +10266,10 @@ Decl *Sema::ActOnParamDeclarator(Scope *S, Declarator &D) {
   // HLSL Change Starts
   if (getLangOpts().HLSL) {
     const bool IsParm = true;
+    if (hlsl::IsStringType(parmDeclType)) {
+      Diag(D.getLocStart(), diag::err_hlsl_unsupported_string_decl) << 1;
```
Comment (tex3d) [inline tools/clang/lib/Sema/SemaDecl.cpp]: Let's comment or create static const named var to indicate what "1" means here.

File: tools/clang/test/HLSL/string.hlsl [hunk 5]
```diff
@@ -1,27 +1,68 @@
...
 
 float4 main() : SV_Target0{
-  float4 cp4_local = cp4;
-  int i = 0;
-  cp4_local.x = a("hi" "bob");
-  printf("hi mom", 1, 2, 3);
-  hello_here("a", "b", 1);
-  return cp4_local;
-}
+  string str;                                               /* expected-error {{string declaration not supported here}} fxc-pass {{}} */
```
Comment (tex3d) [inline tools/clang/test/HLSL/string.hlsl]: add test for struct field too.

File: tools/clang/lib/AST/Expr.cpp (hunk 3 above)
Comment (hekota) [reply tools/clang/lib/AST/Expr.cpp]: This file is mostly original clang code so it is probably best to keep the original assert here (I'll revert the hlsl::InStringLiteralType() check and use the original one). In other cases I'll use DXASSERT where appropriate.

File: tools/clang/test/HLSL/string.hlsl [hunk 6]
```diff
@@ -53,16 +53,20 @@ void hello_here(string message, string s, float f) {        /* expected-error {{
...
   str = s_global2;                                          /* expected-error {{use of undeclared identifier 'str'}} fxc-pass {{}} */
 
-  string strArray[];                                        /* expected-error {{array of type string is not supported}} fxc-error {{X3074: 'strArray': implicit array missing initial value}} */
-  vector<string, 4> strVector1;                             /* expected-error {{'string' cannot be used as a type parameter where a scalar is required}} fxc-error {{X3122: vector element type must be a scalar type}} */
-  matrix<string, 4, 4> strMatrix1;                          /* expected-error {{'string' cannot be used as a type parameter where a scalar is required}} fxc-error {{X3123: matrix element type must be a scalar type}} */
+  string strArray[5];                                        /* expected-error {{array of type string is not supported}} fxc-pass {{}} */
+  vector<string, 4> strVector1;                             /* expected-error {{'string' cannot be used as a type parameter where a scalar is required}} fxc-pass {{}} */
```
Comment (tex3d) [inline tools/clang/test/HLSL/string.hlsl]: How is fxc passing this now, (also the matrix case)?

File: tools/clang/lib/AST/Expr.cpp (hunk 3 above)
Comment (tex3d) [reply tools/clang/lib/AST/Expr.cpp]: Makes sense.

File: tools/clang/test/HLSL/string.hlsl (hunk 6 above)
Comment (hekota) [reply tools/clang/test/HLSL/string.hlsl]: Oops :) There was a semicolon missing at the end of struct decl which it threw fxc off and I did not run VerifierHelper.py after I fixed it. Updating the errors now and also adding a check & test for string as a function return value.

**Pull request merged**